File: Rotate Image.java
public class Solution {
    public void rotate(int[][] matrix) {
        // (x, y) -> (y, W-x) -> (W-x, H-y) -> (H-y, x) -> (x, y)
        int N = matrix.length;
        for (int x = 0; x < (N+1)/2; x++) {
            for (int y = 0; y < N/2; y++) {
                int temp = matrix[x][y];
                matrix[x][y] = matrix[N-y-1][x];
                matrix[N-y-1][x] = matrix[N-1-x][N-1-y];
                matrix[N-1-x][N-1-y] = matrix[y][N-1-x];
                matrix[y][N-1-x] = temp;
            }
        }
    }
}

File: Rotate Image.cpp
class Solution {
public:
    void rotate(vector<vector<int> > &matrix) {
        // (x, y) -> (y, W-x) -> (W-x, H-y) -> (H-y, x) -> (x, y)
        unsigned N = matrix.size();
        for (unsigned i = 0; i < (N + 1) / 2; i++) {
            for (unsigned j = 0; j < N / 2; j++) {
                int temp = matrix[i][j];
                matrix[i][j] = matrix[N-1-j][i];
                matrix[N-1-j][i] = matrix[N-1-i][N-1-j];
                matrix[N-1-i][N-1-j] = matrix[j][N-1-i];
                matrix[j][N-1-i] = temp;
            }
        }
    }
};